### PR TITLE
Add storagemappings to ContainerMeta as ContainerStorageMeta

### LINKED ISSUE #579: ingest mounted location into CharmMeta
Obtaining storage location via the storages section,
```python
self.meta.storages["data"].location
```
is not very useful for k8s charms because for k8s charms that sepcifies mount point in the charm container, not the workload.
For the workload, would need something like:
```python
self.meta.containers[self._container_name].mounts[0].location
```
or
```python
self.meta.containers[self._container_name].mounts.data.location
```

However, mounted location is not ingested by CharmMeta
https://github.com/canonical/operator/blob/b1a8fffc053fbbdaf43b826e78c220bfdae4c8fe/ops/charm.py#L891-L892

### PR DESCRIPTION
**Pull request by rbarry82**
Add a `.mounts` properties to `ContainerMeta`, so charm authors can check:

`self.containers[name].mounts[key].location`

If there are multiple mount points for the same storage, this will raise a `RuntimeError` with a message indicating that the `.locations` property should be used instead, such as:

`self.containers[name].mounts[key].locations`

Closes #579

### CODE REVIEW

File: ops/charm.py [hunk 1]
```diff
@@ -893,7 +893,68 @@ class ContainerMeta:
 
     Attributes:
         name: Name of container (key in the YAML)
+        mounts: :class:`ContainerStorageMeta` mounts available to the container
     """
-
     def __init__(self, name, raw):
         self.name = name
+        self._mounts = {}
```
Comment (sed-i) [inline ops/charm.py:900]:
```suggestion
        self._mounts: typing.Dict[str, ContainerStorageMeta] = {}
```

File: ops/charm.py [hunk 2]
```diff
@@ -893,7 +893,68 @@ class ContainerMeta:
...
+    def _populate_mounts(self, mounts: typing.List):
+        """Populate a list of container mountpoints.
+
+        Since Charm Metadata v2 specifies the mounts as a List, do a little data manipulation
+        to convert the values to "friendly" names which contain a list of mountpoints
+        under each key.
+        """
+        for mount in mounts:
+            storage = mount.get("storage", "")
+            mount = mount.get("location", "")
```
Comment (sed-i) [inline ops/charm.py:920]: Is `""` a safe default? i.e. 
1. is it possible for `if "" in self._mounts` to evaluate to True?
1. would it make sense to `self._mounts[storage] = ContainerStorageMeta(storage, "")`

File: ops/charm.py [hunk 3]
```diff
@@ -893,7 +893,68 @@ class ContainerMeta:
...
+                self._mounts[storage].add_location(mount)
+            else:
+                self._mounts[storage] = ContainerStorageMeta(storage, mount)
+
+
+class ContainerStorageMeta:
+    """Metadata about storage for an individual container.
+
+    Attributes:
+        name: a :class:`StorageMeta` key to mount
```
Comment (sed-i) [inline ops/charm.py:932]:
```suggestion
        storage: a :class:`StorageMeta` key to mount
```
and isn't `storage` a string?

File: ops/charm.py [hunk 4]
```diff
@@ -893,7 +893,68 @@ class ContainerMeta:
...
+    @property
+    def locations(self) -> typing.List:
+        """An accessor for the list of locations for a mount."""
+        return self._locations
+
+    def __getattr__(self, name):
+        if name == "location":
+            if len(self._locations) == 1:
+                return self._locations[0]
+            else:
```
Comment (sed-i) [inline ops/charm.py:952]: is `len == 0` possible?

File: ops/charm.py [hunk 5]
```diff
@@ -893,7 +893,68 @@ class ContainerMeta:
...
+    def add_location(self, location):
+        """Add an additional mountpoint to a known storage."""
+        self._locations.append(location)
+
+    @property
+    def locations(self) -> typing.List:
+        """An accessor for the list of locations for a mount."""
+        return self._locations
+
+    def __getattr__(self, name):
```
Comment (sed-i) [inline ops/charm.py:948]: Why not `@property`?

File: ops/charm.py [hunk 6]
```diff
@@ -893,7 +893,68 @@ class ContainerMeta:
...
+            storage = mount.get("storage", "")
+            mount = mount.get("location", "")
+
+            if storage in self._mounts:
+                self._mounts[storage].add_location(mount)
+            else:
+                self._mounts[storage] = ContainerStorageMeta(storage, mount)
+
+
+class ContainerStorageMeta:
```
Comment (sed-i) [inline ops/charm.py:928]: Why not attribute-like access to dict items?
https://github.com/cdgriffith/Box/wiki
https://github.com/Infinidat/munch

File: test/test_charm.py [hunk 7]
```diff
@@ -430,3 +431,49 @@ def test_containers(self):
...
         self.assertEqual(meta.containers['test1'].name, 'test1')
         self.assertEqual(meta.containers['test2'].name, 'test2')
+
+    def test_containers_storage(self):
+        meta = CharmMeta.from_yaml("""
+name: k8s-charm
+storage:
+  data:
+    type: filesystem
+    location: /test/storage
```
Comment (sed-i) [inline test/test_charm.py:441]: Should probably have the charm location different from the workload location to better test

File: test/test_charm.py [hunk 8]
```diff
@@ -430,3 +431,49 @@ def test_containers(self):
...
+containers:
+  test1:
+    mounts:
+      - storage: data
+        location: /test/storage
+      - storage: data
+        location: /test/other
```
Comment (sed-i) [inline test/test_charm.py:471]: not sure what this non-uniqueness means.
does it mean that the same data found in the charm container under `/test/storage` is mounted twice in the workload container?

File: ops/charm.py (hunk 4 above)
Comment (rbarry82) [reply ops/charm.py:952]: No, but len > 0 is.

File: ops/charm.py (hunk 5 above)
Comment (rbarry82) [reply ops/charm.py:948]: In the case where an identical storage is mounted in multiple places, the `location` property would be incoherent/inconsistent, and dependent on ordering. Not having the property at all in that instance is better.

File: ops/charm.py (hunk 6 above)
Comment (rbarry82) [reply ops/charm.py:928]: This matches existing semantics.

File: test/test_charm.py (hunk 8 above)
Comment (rbarry82) [reply test/test_charm.py:471]: Yes. Specifically, see "Using subPath" in the kubernetes documentation.

The reverse is also possible with projected volumes (primarily for secrets), but I don't think Juju has gotten to that point yet.

File: ops/charm.py [hunk 9]
```diff
@@ -893,7 +893,68 @@ class ContainerMeta:
...
         self.name = name
+        self._mounts = {}
+
+        # This is not guaranteed to be populated/is not enforced yet
+        if raw:
+            self._populate_mounts(raw.get('mounts', []))
+
+    @property
+    def mounts(self) -> typing.Dict:
+        """An accessor for the mounts in a container."""
```
Comment (balbirthomas) [inline ops/charm.py:908]: It may be useful to mention in the docstring what the `dict` keys are and perhaps even values.

File: ops/charm.py [hunk 10]
```diff
@@ -893,7 +893,68 @@ class ContainerMeta:
 
     Attributes:
         name: Name of container (key in the YAML)
+        mounts: :class:`ContainerStorageMeta` mounts available to the container
```
Comment (balbirthomas) [inline ops/charm.py:896]: At what point in the charm lifecycle is `ContainerStorageMeta` populated ? For example is it populated only after the `StorageAttachedEvent`. If so, should this be made explicit in the doc string.

**Review by balbirthomas — APPROVED** [review]

File: ops/charm.py (hunk 10 above)
Comment (rbarry82) [reply ops/charm.py:896]: `ContainerStorageMeta` is populated when the charm metadata is -- i.e. at charm initialization, like the rest of the `*Meta` classes. No dependency on events.

File: ops/charm.py [hunk 11]
```diff
@@ -929,8 +947,14 @@ class ContainerStorageMeta:
     """Metadata about storage for an individual container.
 
     Attributes:
-        name: a :class:`StorageMeta` key to mount
+        storage: a name for the mountpoint, which should exist the keys for :class:`StorageMeta`
+                 for the charm
         location: the location `storage` is mounted at
+        locations: a list of mountpoints for th ekey
```
Comment (petevg) [inline ops/charm.py:953]: Typo. "th ekey" should be "the key"

File: ops/charm.py [hunk 12]
```diff
@@ -893,7 +893,92 @@ class ContainerMeta:
...
+    @property
+    def mounts(self) -> typing.Dict:
+        """An accessor for the mounts in a container.
+
+        Dict keys match key name in :class:`StorageMeta`
+
+        Example::
+
+            storage:
+              FOO:
```
Comment (jnsgruk) [inline ops/charm.py:915]:
```suggestion
              foo:
```

File: ops/charm.py [hunk 13]
```diff
@@ -893,7 +893,92 @@ class ContainerMeta:
...
+                type: filesystem
+                location: /test
+            containers:
+              bar:
+                mounts:
+                  - storage: FOO
```
Comment (jnsgruk) [inline ops/charm.py:921]:
```suggestion
                  - storage: foo
```

File: ops/charm.py [hunk 14]
```diff
@@ -893,7 +893,92 @@ class ContainerMeta:
...
+        if raw:
+            self._populate_mounts(raw.get('mounts', []))
+
+    @property
+    def mounts(self) -> typing.Dict:
+        """An accessor for the mounts in a container.
+
+        Dict keys match key name in :class:`StorageMeta`
+
+        Example::
```
Comment (jnsgruk) [inline ops/charm.py:912]:
```suggestion
        Example:
```

**Review by jnsgruk — DISMISSED** [review]
Nits in docstrings but otherwise LGTM

File: ops/charm.py (hunk 14 above)
Comment (rbarry82) [reply ops/charm.py:912]: This one is actually required by Sphinx+reStructuredText to get a block literal. No single `:` if we want it to render properly

File: test/test_charm.py [hunk 15]
```diff
@@ -430,3 +431,49 @@ def test_containers(self):
...
+  other:
+    type: filesystem
+    location: /test/other
+containers:
+  test1:
+    mounts:
+      - storage: data
+        location: /test/storage
+      - storage: other
+        location: /test/other
```
Comment (sed-i) [inline test/test_charm.py:451]: Having the location here (workload) match exactly the location above (charm) might give false negatives.
Could also add a test for when `location` under `storage` is omitted (should become `/var/lib/juju/storage/<name>/<num>`)

**Review by sed-i — APPROVED** [review]

File: ops/charm.py (hunk 14 above)
Comment (jnsgruk) [reply ops/charm.py:912]: Doh. Brainfart.

File: test/test_charm.py (hunk 15 above)
Comment (jnsgruk) [reply test/test_charm.py:451]: +1 on testing for the default case.

**Review by petevg — APPROVED** [review]
This makes sense to me. Nothing jumps out as difficult to interact with. And most of the complexity is handle on the Juju side -- this is just exposing the data to the charm operator, which is nice.

**Review by jnsgruk — APPROVED** [review]

**Pull request merged**